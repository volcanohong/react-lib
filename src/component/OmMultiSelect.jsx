import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Autocomplete, { createFilterOptions } from '@material-ui/lab/Autocomplete';
import Chip from '@material-ui/core/Chip';
import OmCheckbox from './OmCheckbox';

import CheckBoxOutlineBlankIcon from "@material-ui/icons/CheckBoxOutlineBlank";
import CheckBoxIcon from "@material-ui/icons/CheckBox";
import { Checkbox } from "@material-ui/core";

const useStyles = makeStyles({
  root: {
    '&:focus': {
      outline: 'none',
    }
  }
});

function OmMultiSelect(props) {
  const classes = useStyles();
  const checkboxStyle = {
    marginLeft: -10,
    paddingLeft: 0,
    color: 'grey'
  };
  const listItemStyle = {
    color: 'grey'
  };

  const { className, color, lable, options, width, field, placeholder, limitTags, showTags, onChange, ...others } = props;
  const selectAllLabel = 'Select All';
  const [selectedOptions, setSelectedOptions] = useState(props.value ? props.value : []);
  const allSelected = options.length === selectedOptions?.length;
  const handleSelectAll = isSelected => {
    setSelectedOptions(isSelected ? options : []);
  };

  const handleToggleSelectAll = () => {
    console.log(allSelected);
    handleSelectAll && handleSelectAll(!allSelected);
  };

  const getOptLabel = (opt) => {
    return opt[field] || opt;
  };
  //array of objects can cause unmatch of value and option, which generates lots of warnings
  const handleObjectMatch = (opt, val) => {
    if (field) {
      return opt[field] === val[field];
    }
    return opt === val;
  };

  const handleRenderTags = showTags ? (values, getTagProps) => {
    return values.map((opt, idx) => (
      <Chip {...getTagProps({ idx })} key={idx} label={getOptLabel(opt)} />
    ));
  } : (values, getTagProps) => {
    if (values.length <= limitTags) {
      let tags = values.map((opt, idx) => (
        getOptLabel(opt)
      ));
      return intersep(tags, ", ");
    }
    return values.length + ' Selected';
  };

  const handleChange = (event, selectedOptions, reason) => {
    if (reason === "select-option" || reason === "remove-option") {
      if (selectedOptions.find(opt => getOptLabel(opt) === selectAllLabel)) {
        console.log('1 sel all');
        handleToggleSelectAll();
        let result = options.filter(opt => getOptLabel(opt) !== selectAllLabel);
        // return onChange(result);
        // return result;
      } else {
        console.log('2 sel partial');
        setSelectedOptions(selectedOptions);
        // return onChange(selectedOptions);
        // return selectedOptions;
      }
    } else if (reason === "clear") {
      setSelectedOptions([]);
    }
  };

  const handleRenderOption = (opt, { selected }) => {
    const selectAllProps = getOptLabel(opt) === selectAllLabel ? { checked: allSelected } : {};
    return (
      <React.Fragment>
        <OmCheckbox
          color={color}
          // icon={<CheckBoxOutlineBlankIcon fontSize="small" />}
          // checkedIcon={<CheckBoxIcon fontSize="small" />}
          checked={selected}
          {...selectAllProps}
        />
        <span style={listItemStyle}>{getOptLabel(opt)}</span>
      </React.Fragment>
    );
  };

  const handleRenderInput = (params) => (
    <TextField
      label={lable}
      placeholder={placeholder}
      {...params}
      fullWidth
    />
  );

  const filter = createFilterOptions();
  const handleFilterOptions = (opts, params) => {
    const filtered = filter(opts, params);
    const selectAll = field ? { [field]: selectAllLabel } : selectAllLabel
    return [selectAll, ...filtered];
  }

  return (
    <div className={className}>
      <Autocomplete
        multiple
        disableCloseOnSelect
        limitTags={limitTags}
        onChange={handleChange}
        options={options}
        // value={value}
        value={selectedOptions}
        getOptionLabel={opt => getOptLabel(opt)}
        // getOptionSelected={handleObjectMatch}
        style={{ width: width }}
        ListboxProps={{
          style: {
            maxHeight: 200,
            overflow: 'auto',
            boxShadow: '0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23)'
          }
        }}
        renderTags={handleRenderTags}
        renderOption={handleRenderOption}
        renderInput={handleRenderInput}
        filterOptions={handleFilterOptions}
        {...others}
      />
    </div>
  );
}

function intersep(arr, sep) {
  if (arr.length === 0) {
    return [];
  }
  return arr.slice(1).reduce(function (str, x, i) {
    return str.concat([sep, x]);
  }, [arr[0]]);
}

OmMultiSelect.propTypes = {
  lable: PropTypes.string,
  placeholder: PropTypes.string,
  className: PropTypes.string,
  color: PropTypes.string,
  options: PropTypes.array,
  // value: PropTypes.array,
  field: PropTypes.string,
  width: PropTypes.number,
  limitTags: PropTypes.number,
  showTags: PropTypes.bool,
  diabled: PropTypes.bool,
  onChange: PropTypes.func
};

OmMultiSelect.defaultProps = {
  options: [],
  // value: [],
  color: 'primary',
  width: 300,
  limitTags: 1,
  showTags: true,
  // onChange: () => { }
};

export default React.memo(OmMultiSelect);