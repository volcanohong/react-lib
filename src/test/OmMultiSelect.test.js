import React from 'react';
import { create, fireEvent, waitForElement } from 'react-test-renderer';
import OmMultiSelect from '../component/OmMultiSelect';

describe('OmMultiSelect Component', () => {
    const mockedOptions = [
        'Option 1',
        'Option 2',
        'Option 3',
        'Option 4',
        'Option 5',
        'Option 6'
    ];
    it('it should render a multi-select', () => {
        const select = create(<OmMultiSelect options={mockedOptions} />);
        expect(select.toJSON()).toMatchSnapshot();
    });

    it("it should have props", () => {
        const select = create(<OmMultiSelect options={[]} label='Lable' placeholder='Placeholder' />).root;
        expect(select.props.options).toEqual([]);
        expect(select.props.placeholder).toEqual('Placeholder');
    });



    // it("it should have options and value", () => {
    //     const mockedOnChange = jest.fn();
    //     const select = create(<OmMultiSelect options={mockedOptions} value={mockedOptions[0]} onChange={mockedOnChange} />).root;
    //     expect(select.props.options).toEqual(mockedOptions);
    //     expect(select.props.value).toEqual(mockedOptions[0]);

    //     const selectComponent = select.children[0];
    //     expect(selectComponent).toBeDefined();
    //     expect(selectComponent).not.toBeNull();
    //     expect(mockedOnChange).toHaveBeenCalledTimes(0);
    // });
});