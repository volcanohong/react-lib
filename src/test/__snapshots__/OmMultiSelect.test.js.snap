// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OmMultiSelect Component it should render a multi-select 1`] = `
<div>
  <div
    aria-expanded={false}
    aria-owns={null}
    className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="combobox"
    style={
      Object {
        "width": 300,
      }
    }
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        onClick={[Function]}
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-invalid={false}
          autoCapitalize="none"
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          required={false}
          spellCheck="false"
          type="text"
          value=""
        />
        <div
          className="MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Clear"
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="Open"
            className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Open"
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
